types/+MPCT/code_MPCT_ADMM_semiband_C.c

```diff
@@ -0,0 +1,930 @@
+                }
+                
+            }
+
+        }
+
+        for (unsigned int i = nn_ ; i < nm_ ; i++){
+
+            for (unsigned int l = 0 ; l < NN_+1 ; l++){
+            
+                for (unsigned int j = 0 ; j < mm_ ; j++){
+                    if (l == NN_){
+                        z2[i] += M_hat_u[i-nn_][j+mm_]*z1_ac[l*nm_+nn_+j];
+                    }
+                    else{
+                        z2[i] += M_hat_u[i-nn_][j]*z1_ac[l*nm_+nn_+j];
+                    }
+                }
+
+            }
+
+        }
+
+        for (unsigned int i = nm_ ; i < 2*nn_+mm_ ; i++){
+
+            for (unsigned int l = 0 ; l < NN_+1 ; l++){
+
+                for (unsigned int j = 0 ; j < nn_ ; j++){
+
+                    if(l == NN_){
+                        z2[i] += M_hat_x[i-mm_][j+nn_]*z1_ac[l*nm_+j];
+                    }
+                    else{
+                        z2[i] += M_hat_x[i-mm_][j]*z1_ac[l*nm_+j];
+                    }
+
+                    
+                }
+
+            }
+
+        }
+
+        for (unsigned int i = nm_+nn_ ; i < 2*nm_ ; i++){
+
+            for (unsigned int l = 0 ; l < NN_+1 ; l++){
+
+                for (unsigned int j = 0 ; j < mm_ ; j++){
+                    
+                    if (l == NN_){
+                        z2[i] += M_hat_u[i-2*nn_][j+mm_]*z1_ac[l*nm_+nn_+j];
+                    }
+                    else{
+                        z2[i] += M_hat_u[i-2*nn_][j]*z1_ac[l*nm_+nn_+j];
+                    }
+
+                }
+
+            }
+
+        }
+
+        memset(aux_0, 0, sizeof(double)*(NN_+1)*nm_);
+
+        // (U_hat * z2_a) computed sparsely
+        for (unsigned int l = 0 ; l < NN_ ; l++){
+
+            for (unsigned int i = l*nm_ ; i < l*nm_ + nn_ ; i++){
+
+                for (unsigned int j = 0 ; j < nn_ ; j++){
+
+                    aux_0[i] -= Q[i-l*nm_][j] * z2[j];
+
+                }
+
+            }
+
+            for (unsigned int  i = l*nm_+nn_ ; i < (l+1)*nm_ ; i++){
+
+                for (unsigned int j = 0 ; j < mm_ ; j++){
+
+                    aux_0[i] -= R[i-l*nm_-nn_][j] * z2[j+nn_];
+
+                }
+
+            }
+
+        }
+
+        for (unsigned int i = NN_*nm_ ; i < (NN_+1)*nm_ ; i++){
+
+            aux_0[i] = z2[i-(NN_-1)*nm_];
+
+        }
+        // End of computation of (U_hat * z2_a)
+
+        solve_banded_diag_sys(Q_rho_i, R_rho_i, S_rho_i, T_rho_i, z3_ac, aux_0); // Obtains z3_a
+
+        // Computation of xi, which is the solution of eq. (9a)
+        for (unsigned int i = 0 ; i < (NN_+1)*nm_ ; i++){
+
+            xi[i] = z1_ac[i] - z3_ac[i];
+
+        }
+        // End of computation of xi
+
+
+        /****** Compute mu from eq. (9b) using Alg. 2 from the article ******/
+        
+        // Sparse computation of -(G*xi+b)
+        memset(aux_1, 0, sizeof(double)*(NN_+2)*nn_);
+
+        for (unsigned int i = 0 ; i < nn_ ; i++){
+
+            aux_1[i] = -(b[i]+xi[i]);
```

Comment:
Array `b` could also be eliminated. I think it is only used here, where we could instead directly use `x0[i]`.

Reply:
Done